### PR TITLE
Initial draft of Cluster user guide

### PR DESCRIPTION
**Pull request by writinwaters**
### Type of change

- [x] Documentation Update

### CODE REVIEW

File: docs/guides/set_up_cluster.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+### Log synchronization
+
+### Configurations
+
+#### Cluster-specific settings
+
+#### References
+
+- [infinity/conf/leader.toml](https://github.com/infiniflow/infinity/blob/main/conf/leader.toml)
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:46]: Three roles shouldn't have difference in config file.

File: docs/guides/set_up_cluster.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+- **Follower node**: The synchronous log/WAL write node for the leader. It acts as a backup for the leader node, maintaining strong consistency with its state. 
+- **Learner node**: The *asynchronous* log/WAL write node for the leader. A learner also serves as a backup for the leader node. However, Its state may be behind that of the leader, as it is not required to maintain strong consistency with the leader, and the leader does not need to confirm whether all learner nodes have completed log persistence.
+
+### Startup and communication processes
+
+When started up in cluster mode, a node enters `ADMIN` state, but is not automatically assigned a role like leader, follower, or learner. You must call `ADMIN SET NODE ROLE` to assign it a role. Once a leader node starts, it does the following:
+
+1. Reads logs from the local disk to determine the metadata and data to read from shared storage.
+2. Waits for connections from follower or learner nodes.
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:35]: Remove 'waits for connections'

File: docs/guides/set_up_cluster.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+Once you set a node to follower or learner, it registers with the leader node as per the IP address specified in `ADMIN SET NODE ROLE`. Upon receiving the registration request, the leader node sends back its current log and snapshot locations for the followers and learners to restore their database state from shared storage.
```
[the remark this answers is not in the document]
Comment (JinHai-CN) [reply docs/guides/set_up_cluster.md:37]: IP address -> address

[the remark this answers is not in the document]
Comment (JinHai-CN) [reply docs/guides/set_up_cluster.md:37]: Remove 'and snapshot locations for'

[the remark this answers is not in the document]
Comment (JinHai-CN) [reply docs/guides/set_up_cluster.md:37]: Once you set a node to follower or learner, it registers to the leader node using command `ADMIN SET NODE ROLE`. Upon receiving the registration request, the leader node sends back its log for the followers and learners to construct their database state from shared storage.

File: docs/guides/set_up_cluster.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+### Keep-alive mechanism
+
+Once successfully registered with the leader node, a follower or learner starts sending periodic heartbeats to it. The leader node relies on these heartbeats to manage the online status of each node. For example, if it does not receive heartbeats from a particular node for a specified time period, it sets that node's online status to `timeout`.
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:41]: online status -> connection status

File: docs/guides/set_up_cluster.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+### Log synchronization
+
+When a transaction occurs, the leader node sends its log to both follower and learner nodes. The leader confirms the transaction's completion only after receiving confirmation that all its follower nodes have successfully persisted the log. While the leader also sends logs to learner nodes, it does not require confirmation from them.
+
+### Configurations
+
+#### Cluster-specific settings
+
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:50]: Only need to mention:
server_mode = "admin"
storage_type = "minio"

File: docs/guides/set_up_cluster.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (writinwaters) [reply docs/guides/set_up_cluster.md:37]: register with

File: docs/guides/set_up_cluster.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+````shell
+```shell
+curl --request POST \
+    --url http://localhost:23821/admin/node/current \
+    --header 'accept: application/json' \
+    --header 'content-type: application/json' \
+    --data ' {
+        "role" : "leader",
+        "name" : "Trump",
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:105]: Trump -> Captain

File: docs/guides/set_up_cluster.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+````shell
+```shell
+curl --request POST \
+    --url http://localhost:23821/admin/node/current \
+    --header 'accept: application/json' \
+    --header 'content-type: application/json' \
+    --data ' {
+        "role" : "leader",
+        "name" : "HarryPotter",
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:94]: Harry

File: docs/guides/set_up_cluster.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+
+![infinity_cluster](https://github.com/user-attachments/assets/3e9abeed-1698-4741-8bdb-ba3b05c1d7a3)
+
+### Architecture
+
+Infinity employs a distributed architecture comprising one leader node and *N* follower nodes, where 0 &le; *N* &le; 4. As illustrated in the diagram above, all nodes in the cluster use MinIO for persistent storage.
+
+- **Leader node**: The node responsible for processing transactions and managing connection status of other nodes in the cluster. When a transaction occurs, the leader node transmits the logs to both follower and learner nodes. The leader node confirms the completion of the transaction only upon receiving messages confirming completion of log persistence from *all* follower nodes.
+- **Follower node**: Receives log/WAL from the leader synchronously. It acts as a backup for the leader node, maintaining strong consistency with the leader's state.
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:27]: leader's state -> leader's data state

File: docs/guides/set_up_cluster.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+- **Learner node**: Receives log/WAL from the leader *asynchronously*. A learner also serves as a backup for the leader node. However, its state may be behind that of the leader, because it is not required to maintain strong consistency with the leader, and neither does the leader need to confirm whether all learner nodes have completed log persistence.
+
+From the user's perspective, the leader is the only write node, and all write operations must go through the leader node; all nodes in the cluster serve as read nodes, allowing you to send read operations to any of the leader, follower, or learner nodes, thereby alleviating the write burden on the leader.
+
+### Startup and communication processes
+
+When started up in cluster mode, a node enters `ADMIN` state, but is not automatically assigned a role like leader, follower, or learner. You must call `ADMIN SET NODE ROLE` to assign it a role. Once a leader node starts, it reads logs from the local disk to determine the metadata and data to read from shared storage.
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:34]: `ADMIN` mode

File: docs/guides/set_up_cluster.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    --header 'accept: application/json' \
+    --header 'content-type: application/json' \
+    --data ' {
+        "role" : "leader",
+        "name" : "Trump",
+    } '
+```
+````
+
+*When the method call succeeds, the node's state changes to `LEADER`.*
```
[the remark this answers is not in the document]
Comment (writinwaters) [reply docs/guides/set_up_cluster.md]: Added ADMIN LIST NODES instead.

File: docs/guides/set_up_cluster.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    --url http://localhost:23821/admin/node/current \
+    --header 'accept: application/json' \
+    --header 'content-type: application/json' \
+    --data ' {
+        "role" : "leader",
+        "name" : "Harry",
+    } '
+```
+````
+
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:86]: User can use following command to check current node role and status:

curl --request GET \
    --url http://localhost:23822/admin/node/current \
    --header 'accept: application/json'

File: docs/guides/set_up_cluster.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        "name" : "Hermione",
+        "address" : "0.0.0.0:23851"
+    } '
+```
+
+*When the method call succeeds, the node is promoted to follower and registered with the leader node, which listens on `0.0.0.0:23851`.*
+
+:::tip NOTE
+
+A node in `ADMIN` mode with `storage_type = "minio"` or in `CLUSTER` mode (as a learner node) can be promoted to follower node.
```
Comment (JinHai-CN) [inline docs/guides/set_up_cluster.md:115]: Why learner node doesn't has this part?

**Review by JinHai-CN — APPROVED** [review]

**Pull request merged**